File: apps/app/convex/_lib/breadth.ts
/**
 * Full-watchlist breadth aggregates.
 *
 * Lives in convex/_lib so Convex functions can use it without importing from
 * the Next.js app tree (Convex bundles independently via esbuild — app-side
 * imports like `next/headers` or `server-only` would break deployment).
 * The app re-exports these from `src/lib/overview-daily-brief.ts`.
 */

/**
 * Aggregates computed over the FULL watchlist (not just ranked movers) at
 * snapshot build time, so the brief can describe breadth across 70+ names
 * instead of extrapolating from the top handful.
 */
export type BreadthStats = {
  advancers: number;
  decliners: number;
  flat: number;
  medianChangePct: number;
  /** p90 − p10 of change percents — a real dispersion measure. */
  spreadPct: number;
  /** Names that moved at least ±5%. */
  bigMovers: number;
};

const BREADTH_FLAT_BAND_PCT = 0.5;
const BREADTH_BIG_MOVE_PCT = 5;

export function computeBreadthStats(changePcts: ReadonlyArray<number>): BreadthStats | null {
  const values = changePcts.filter((x) => typeof x === "number" && Number.isFinite(x));
  if (values.length === 0) return null;

  const sorted = [...values].sort((a, b) => a - b);
  const quantile = (p: number): number => {
    const idx = (sorted.length - 1) * p;
    const lo = Math.floor(idx);
    const hi = Math.ceil(idx);
    const loVal = sorted[lo] ?? 0;
    const hiVal = sorted[hi] ?? loVal;
    return loVal + (hiVal - loVal) * (idx - lo);
  };

  let advancers = 0;
  let decliners = 0;
  let flat = 0;
  for (const x of values) {
    if (x > BREADTH_FLAT_BAND_PCT) advancers++;
    else if (x < -BREADTH_FLAT_BAND_PCT) decliners++;
    else flat++;
  }

  return {
    advancers,
    decliners,
    flat,
    medianChangePct: quantile(0.5),
    spreadPct: quantile(0.9) - quantile(0.1),
    bigMovers: values.filter((x) => Math.abs(x) >= BREADTH_BIG_MOVE_PCT).length,
  };
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Overview/Breadth.swift
import Foundation

/// Port of `convex/_lib/breadth.ts`.
public struct BreadthStats: Sendable, Hashable, Codable {
  public var advancers: Int
  public var decliners: Int
  public var flat: Int
  public var medianChangePct: Double
  public var spreadPct: Double
  public var bigMovers: Int

  public init(advancers: Int, decliners: Int, flat: Int, medianChangePct: Double, spreadPct: Double, bigMovers: Int) {
    self.advancers = advancers; self.decliners = decliners; self.flat = flat
    self.medianChangePct = medianChangePct; self.spreadPct = spreadPct; self.bigMovers = bigMovers
  }

  enum CodingKeys: String, CodingKey { case advancers, decliners, flat, medianChangePct, spreadPct, bigMovers }

  public init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: CodingKeys.self)
    advancers = Int(try c.decode(Double.self, forKey: .advancers))
    decliners = Int(try c.decode(Double.self, forKey: .decliners))
    flat = Int(try c.decode(Double.self, forKey: .flat))
    medianChangePct = try c.decode(Double.self, forKey: .medianChangePct)
    spreadPct = try c.decode(Double.self, forKey: .spreadPct)
    bigMovers = Int(try c.decode(Double.self, forKey: .bigMovers))
  }

  public var total: Int { advancers + decliners + flat }

  static let flatBandPct = 0.5
  static let bigMovePct = 5.0

  public static func compute(_ changePcts: [Double]) -> BreadthStats? {
    let values = changePcts.filter { $0.isFinite }
    guard !values.isEmpty else { return nil }
    let sorted = values.sorted()
    func quantile(_ p: Double) -> Double {
      let idx = Double(sorted.count - 1) * p
      let lo = Int(idx.rounded(.down)), hi = Int(idx.rounded(.up))
      let loVal = sorted[lo], hiVal = sorted[hi]
      return loVal + (hiVal - loVal) * (idx - Double(lo))
    }
    var adv = 0, dec = 0, flat = 0
    for x in values {
      if x > flatBandPct { adv += 1 } else if x < -flatBandPct { dec += 1 } else { flat += 1 }
    }
    return BreadthStats(advancers: adv, decliners: dec, flat: flat, medianChangePct: quantile(0.5),
                        spreadPct: quantile(0.9) - quantile(0.1), bigMovers: values.filter { abs($0) >= bigMovePct }.count)
  }
}
